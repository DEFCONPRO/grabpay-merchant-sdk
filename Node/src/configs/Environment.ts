import { TCountry, TEnvironment } from '../types';

export type EnvironmentSet = {
  baseUrl: string;
};

const definedEnvs: Partial<Record<TCountry, Record<TEnvironment, EnvironmentSet>>> = {
  VN: {
    STG: {
      baseUrl: 'https://stg-paysi.moca.vn',
    },
    PRD: {
      baseUrl: 'https://partner-gw.moca.vn',
    },
  },
};

const fallbackEnvironment: Record<TEnvironment, EnvironmentSet> = {
  STG: {
    baseUrl: 'https://partner-api.stg-myteksi.com',
  },
  PRD: {
    baseUrl: 'https://partner-api.grab.com',
  },
};

export function getEnvs(country: TCountry, env: TEnvironment) {
  const countryEnv = country in definedEnvs ? definedEnvs[country] : undefined;
  return countryEnv ? countryEnv[env] : fallbackEnvironment[env];
}
