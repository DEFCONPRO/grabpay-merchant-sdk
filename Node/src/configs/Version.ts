export const SDK_VERSION = '2.0.0';
export const SDK_SIGNATURE = '6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7';
