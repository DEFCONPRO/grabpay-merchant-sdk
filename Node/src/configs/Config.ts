import { TCountry, TEnvironment } from '../types';
import { ApiUrlSet, getUrls } from './ApiUrls';
import { EnvironmentSet, getEnvs } from './Environment';

const definedCountryScopes: Partial<Record<TCountry, string>> = {
  VN: 'payment.vn.one_time_charge',
};
const fallbackScope = ['openid', 'payment.one_time_charge'].join(' ');

type CommonClientRawConfig = {
  env: TEnvironment;
  country: TCountry;
  partnerId: string;
  partnerSecret: string;
  merchantId: string;
};

export type OfflineClientRawConfig = CommonClientRawConfig & {
  terminalId: string;
};

export type OnlineClientRawConfig = CommonClientRawConfig & {
  clientId: string;
  clientSecret: string;
  redirectUrl: string;
};

export type ClientRawConfig = OnlineClientRawConfig | OfflineClientRawConfig;

export type InitializedConfig<T extends ClientRawConfig> = {
  envVars: EnvironmentSet;
  urlSet: ApiUrlSet;
  scope: string;
} & T;

export type OnlineClientConfig = InitializedConfig<OnlineClientRawConfig>;
export type OfflineClientConfig = InitializedConfig<OfflineClientRawConfig>;
export type ClientConfig = OnlineClientConfig | OfflineClientConfig;

export function initializeWithConfig<T extends ClientRawConfig>(input: T): InitializedConfig<T> {
  return {
    ...input,
    envVars: getEnvs(input.country, input.env),
    urlSet: getUrls(input.country),
    scope: (input.country in definedCountryScopes ? definedCountryScopes[input.country] : undefined) || fallbackScope,
  };
}
