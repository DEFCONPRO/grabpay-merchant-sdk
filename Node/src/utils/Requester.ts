import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import { generateHmac, getGMTString } from './Utils';
import { SDK_SIGNATURE, SDK_VERSION } from '../configs/Version';
import { ClientConfig } from '../configs/Config';
import { THttpMethod } from '../types';

export interface IRequester {
  hmacRequest<TResponseBody, TRequestBody>(
    request: Readonly<RequestParams<TRequestBody>>,
  ): Promise<Response<TResponseBody>>;
  request<TResponseBody, TRequestBody>(
    request: Readonly<RequestParams<TRequestBody>>,
  ): Promise<Response<TResponseBody>>;
}

export type RequestParams<TRequestBody> = {
  path: string;
  method: THttpMethod;
  query?: Record<string, string>;
  headers?: Record<string, string>;
} & ({ body: string } | { data: TRequestBody } | {});

export type Response<TResponseBody> = {
  status: number;
  headers: Record<string, string>;
  data: TResponseBody;
};

function buildHeadersForHmacRequest<TRequestBody>(
  requestParams: Readonly<RequestParams<TRequestBody>>,
  partnerId: string,
  partnerSecret: string,
): Record<string, string> {
  const { headers = {}, path, method, query } = requestParams;
  const contentType = headers['Content-Type'] || 'application/json';
  const timestamp = headers['Date'] || getGMTString();
  let pathWithParams = path;
  if (query && Object.keys(query).length > 0) {
    pathWithParams = `${pathWithParams}?${new URLSearchParams(query).toString()}`;
  }
  const hmac = generateHmac({
    contentType,
    httpMethod: method || 'GET',
    requestBody: 'body' in requestParams ? requestParams.body : '',
    requestUrl: pathWithParams,
    secret: partnerSecret,
    timestamp,
  });
  return {
    Authorization: `${partnerId}:${hmac}`,
    'Content-Type': contentType,
    Date: timestamp,
    ...headers,
  };
}

export default class AxiosRequester implements IRequester {
  _config: ClientConfig;
  _requester: AxiosInstance;

  constructor(config: ClientConfig) {
    const { baseUrl } = config.envVars;
    this._requester = axios.create({ baseURL: baseUrl, transformRequest: [] });
    this._config = config;
  }

  hmacRequest<TResponseBody, TRequestBody>(requestParams: Readonly<RequestParams<TRequestBody>>) {
    const { path, method, query } = requestParams;
    let body = '';
    if ('body' in requestParams) {
      body = requestParams.body;
    } else if ('data' in requestParams) {
      body = JSON.stringify(requestParams.data);
    }
    const hmacSignedHeaders = buildHeadersForHmacRequest(
      {
        ...requestParams,
        body,
      },
      this._config.partnerId,
      this._config.partnerSecret,
    );

    return this.request<TResponseBody, never>({
      headers: hmacSignedHeaders,
      body,
      method,
      path,
      query,
    });
  }

  async request<TResponseBody, TRequestBody>(
    requestParams: Readonly<RequestParams<TRequestBody>>,
  ): Promise<Response<TResponseBody>> {
    let body = '';
    if ('body' in requestParams) {
      body = requestParams.body;
    } else if ('data' in requestParams) {
      body = JSON.stringify(requestParams.data);
    }
    const axiosConfig: AxiosRequestConfig = {
      method: requestParams.method,
      url: requestParams.path,
      headers: {
        ...requestParams.headers,
        'X-Sdk-Country': this._config.country,
        'X-Sdk-Language': 'nodejs',
        'X-Sdk-Signature': SDK_SIGNATURE,
        'X-Sdk-Version': SDK_VERSION,
      },
      ...(requestParams.method === 'GET' ? {} : { data: body }),
      ...(requestParams.query && Object.keys(requestParams.query).length > 0 ? { params: requestParams.query } : {}),
    };

    try {
      const response = await this._requester.request<TResponseBody>(axiosConfig);
      return {
        status: response.status,
        headers: response.headers,
        data: response.data,
      };
    } catch (err) {
      if (axios.isAxiosError(err) && err.response) {
        return {
          status: err.response.status,
          headers: err.response.headers,
          data: err.response.data,
        };
      }
      throw err;
    }
  }
}
