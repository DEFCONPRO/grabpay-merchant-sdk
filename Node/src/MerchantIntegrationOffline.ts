import Requester, { IRequester } from './utils/Requester';
import { initializeWithConfig, OfflineClientConfig } from './configs/Config';
import { SDK_VERSION } from './configs/Version';
import { OfflineClientRawConfig } from './configs/Config';
import * as Utils from './utils/Utils';
import { TCurrency } from './types';

export type CreateQrCodeParams = {
  msgID?: string;
  partnerTxID: string;
  amount: number;
  currency: TCurrency;
};

export type CreateQrCodeResponse = {
  msgID: string;
  qrcode: string;
  txID: string;
};

export type PerformQrCodeTxnParams = {
  msgID?: string;
  partnerTxID: string;
  amount: number;
  currency: TCurrency;
  code: string;
  additionalInfo?: 'amountBreakdown'[];
};

export type PerformQrCodeTxnResponse = {
  msgID: string;
  txID: string;
  status: 'success' | 'failed' | 'unknown' | 'pending' | 'bad_debt';
  currency: TCurrency;
  amount: number;
  updated: number;
  errMsg: string;
  additionalInfo?: {
    amountBreakdown?: Record<string, number | string>;
  };
};

export type CancelTxnParams = {
  msgID?: string;
  partnerTxID: string;
  origPartnerTxID: string;
  origTxID: string;
  currency: TCurrency;
};

export type CancelTxnResponse = {};

export type RefundTxnParams = {
  msgID?: string;
  partnerTxID: string;
  amount: number;
  currency: TCurrency;
  origPartnerTxID: string;
  reason: string;
};

export type RefundTxnResponse = {
  msgID: string;
  txID: string;
  originTxID: string;
  status: 'success' | 'failed' | 'unknown' | 'pending' | 'bad_debt';
  description: string;
  additionalInfo?: {
    amountBreakdown?: Record<string, number | string>;
  };
};

export type GetTxnDetailsParams = {
  msgID?: string;
  partnerTxID: string;
  currency: string;
};

export type GetTxnDetailsResponse = {
  msgID: string;
  txID: string;
  status: 'success' | 'failed' | 'unknown' | 'pending' | 'bad_debt';
  currency: TCurrency;
  amount: number;
  updated: number;
  errMsg: string;
  additionalInfo?: {
    amountBreakdown?: Record<string, number | string>;
  };
};

export class MerchantIntegrationOffline {
  _requester: Requester;
  _config: OfflineClientConfig;

  /**
   * Constructor
   */
  constructor(rawConfig: OfflineClientRawConfig, requester: Requester | null = null) {
    this._config = initializeWithConfig(rawConfig);
    this._requester = requester ?? new Requester(this._config);
  }

  /**
   * Create QR Code
   */
  createQrCode(payload: CreateQrCodeParams) {
    const data = configurePayload(payload, this._config);
    return this._requester.hmacRequest<CreateQrCodeResponse, typeof data>({
      path: this._config.urlSet.POS_CREATE_QR_CODE,
      method: 'POST',
      data,
    });
  }

  /**
   * Perform a Transaction
   */
  performQrCode(payload: PerformQrCodeTxnParams) {
    const data = configurePayload(payload, this._config);
    return this._requester.hmacRequest<PerformQrCodeTxnResponse, typeof data>({
      path: this._config.urlSet.POS_PERFORM_TRANSACTION,
      method: 'POST',
      data,
    });
  }

  /**
   * Cancel a Transaction
   */
  cancel({ origPartnerTxID, ...payload }: CancelTxnParams) {
    const data = configurePayload(payload, this._config);
    const path = Utils.replaceUrl(this._config.urlSet.POS_CANCEL_TRANSACTION, {
      origPartnerTxID,
    });

    return this._requester.hmacRequest<CancelTxnResponse, typeof data>({
      path,
      method: 'PUT',
      data,
    });
  }

  /**
   * Refund a POS Payment
   */
  refund({ origPartnerTxID, ...payload }: RefundTxnParams) {
    const data = configurePayload(payload, this._config);
    const path = Utils.replaceUrl(this._config.urlSet.POS_REFUND_TRANSACTION, {
      origPartnerTxID,
    });

    return this._requester.hmacRequest<RefundTxnResponse, typeof data>({
      path,
      method: 'PUT',
      data,
    });
  }

  /**
   * Check Transaction Details
   */
  getTxnDetails({ partnerTxID, ...payload }: GetTxnDetailsParams) {
    const configuredPayload = {
      ...configurePayload(payload, this._config),
      txType: 'P2M',
    };
    const path = Utils.replaceUrl(this._config.urlSet.POS_GET_TXN_DETAIL, {
      partnerTxID,
    });

    return this._requester.hmacRequest<GetTxnDetailsResponse, never>({
      path,
      method: 'GET',
      query: configuredPayload,
    });
  }

  /**
   * Check Refund Details
   */
  getRefundDetails({ partnerTxID, ...payload }: GetTxnDetailsParams) {
    const configuredPayload = {
      ...configurePayload(payload, this._config),
      txType: 'Refund',
    };
    const path = Utils.replaceUrl(this._config.urlSet.POS_GET_TXN_DETAIL, {
      partnerTxID,
    });

    return this._requester.hmacRequest<CancelTxnResponse, never>({
      path,
      method: 'GET',
      query: configuredPayload,
    });
  }

  /**
   * Get SDK Version
   *
   * @returns {string}
   */
  version() {
    return SDK_VERSION;
  }
}

const configurePayload = <T extends unknown & { msgID?: string }>(
  { msgID, ...payload }: T,
  config: OfflineClientConfig,
) => ({
  ...payload,
  // we use the provided msgID if mex provided one, otherwise we generate one for the request
  msgID: msgID ? msgID : Utils.generateMsgId(),
  grabID: config.merchantId,
  terminalID: config.terminalId,
});
