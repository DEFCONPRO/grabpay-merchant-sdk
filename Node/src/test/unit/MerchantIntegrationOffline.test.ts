import { initializeWithConfig, OfflineClientRawConfig } from '../../configs/Config';
import Requester from '../../utils/Requester';
import * as Utils from '../../utils/Utils';
import { MerchantIntegrationOffline } from '../../MerchantIntegrationOffline';

const partnerTxID = 'ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k'; // generated with `Utils.generateRandomString(32)`
const origPartnerTxID = 'fBTcFk0T1OekPrLtbQoiIB33TyzwrWZl'; // generated with `Utils.generateRandomString(32)`

describe('MerchantIntegrationOffline', () => {
  const config: OfflineClientRawConfig = {
    env: 'STG',
    country: 'VN',
    partnerId: 'partner-id',
    partnerSecret: 'partner-secret',
    merchantId: 'merchant-id',
    terminalId: 'terminal-id',
  };
  let client: MerchantIntegrationOffline;
  const requester = new Requester(initializeWithConfig(config));
  const axiosRequestSpy = jest
    .spyOn(requester._requester, 'request')
    .mockReturnValue(Promise.resolve({ status: 200, headers: {}, data: {} }));

  beforeEach(() => {
    client = new MerchantIntegrationOffline(config, requester);
    jest.useFakeTimers().setSystemTime(0);
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  describe('createQrCode', () => {
    it('use generated msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.createQrCode({ partnerTxID, amount: 100000, currency: 'VND' });
      expect(generateMsgIdSpy).toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
    it('use provided msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.createQrCode({ msgID: 'provided-msgID', partnerTxID, amount: 100000, currency: 'VND' });
      expect(generateMsgIdSpy).not.toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe('performQrCode', () => {
    it('use generated msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.performQrCode({ partnerTxID, amount: 100000, currency: 'VND', code: 'code' });
      expect(generateMsgIdSpy).toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
    it('use provided msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.performQrCode({
        msgID: 'provided-msgID',
        partnerTxID,
        amount: 100000,
        currency: 'VND',
        code: 'code',
      });
      expect(generateMsgIdSpy).not.toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe('cancel', () => {
    it('use generated msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.cancel({ partnerTxID, origPartnerTxID, origTxID: 'txID', currency: 'VND' });
      expect(generateMsgIdSpy).toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
    it('use provided msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.cancel({
        msgID: 'provided-msgID',
        partnerTxID,
        origPartnerTxID,
        origTxID: 'txID',
        currency: 'VND',
      });
      expect(generateMsgIdSpy).not.toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe('refund', () => {
    it('use generated msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.refund({ partnerTxID, origPartnerTxID, amount: 10000, currency: 'VND', reason: 'refund' });
      expect(generateMsgIdSpy).toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
    it('use provided msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.refund({
        msgID: 'provided-msgID',
        amount: 10000,
        partnerTxID,
        origPartnerTxID,
        currency: 'VND',
        reason: 'refund',
      });
      expect(generateMsgIdSpy).not.toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe('getTxnDetails', () => {
    it('use generated msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.getTxnDetails({ partnerTxID, currency: 'VND' });
      expect(generateMsgIdSpy).toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
    it('use provided msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.getTxnDetails({
        msgID: 'provided-msgID',
        partnerTxID,
        currency: 'VND',
      });
      expect(generateMsgIdSpy).not.toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe('getRefundDetails', () => {
    it('use generated msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.getRefundDetails({ partnerTxID, currency: 'VND' });
      expect(generateMsgIdSpy).toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
    it('use provided msgID', async () => {
      const generateMsgIdSpy = jest.spyOn(Utils, 'generateMsgId').mockReturnValue('generated-msgID');
      await client.getRefundDetails({
        msgID: 'provided-msgID',
        partnerTxID,
        currency: 'VND',
      });
      expect(generateMsgIdSpy).not.toHaveBeenCalled();
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });
});
