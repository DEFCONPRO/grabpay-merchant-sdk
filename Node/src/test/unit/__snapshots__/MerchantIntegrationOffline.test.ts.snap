// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerchantIntegrationOffline cancel use generated msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"origTxID\\":\\"txID\\",\\"currency\\":\\"VND\\",\\"msgID\\":\\"generated-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:OPRHW3x99MJkMKwtmgl273yjJ8a/E+85WwfVVRyQfUs=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "PUT",
      "url": "/mocapay/partners/v1/terminal/transaction/fBTcFk0T1OekPrLtbQoiIB33TyzwrWZl/cancel",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline cancel use provided msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"origTxID\\":\\"txID\\",\\"currency\\":\\"VND\\",\\"msgID\\":\\"provided-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:40+XN6V8sXvUgiONqvlbSUotUZf9uBiwXZ1h96ttgh4=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "PUT",
      "url": "/mocapay/partners/v1/terminal/transaction/fBTcFk0T1OekPrLtbQoiIB33TyzwrWZl/cancel",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline createQrCode use generated msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"msgID\\":\\"generated-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:N10r30Jwv6OUWEIo81IiyNWYUnjlWazBhf2vjJI5wOg=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partners/v1/terminal/qrcode/create",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline createQrCode use provided msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"msgID\\":\\"provided-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:CBsc+9xiXgpdlT5n9KFyKHvfaPhqwSvk2M6xaWbimGU=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partners/v1/terminal/qrcode/create",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline getRefundDetails use generated msgID 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "partner-id:UxYDKjb5AirCkx6bnmua15XutXVwuUfj/brU0iBiOzI=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "GET",
      "params": Object {
        "currency": "VND",
        "grabID": "merchant-id",
        "msgID": "generated-msgID",
        "terminalID": "terminal-id",
        "txType": "Refund",
      },
      "url": "/mocapay/partners/v1/terminal/transaction/ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline getRefundDetails use provided msgID 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "partner-id:+Ey26YPUFirhI62Lew1Q9NgALzS10rZrei0CDKrtcXU=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "GET",
      "params": Object {
        "currency": "VND",
        "grabID": "merchant-id",
        "msgID": "provided-msgID",
        "terminalID": "terminal-id",
        "txType": "Refund",
      },
      "url": "/mocapay/partners/v1/terminal/transaction/ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline getTxnDetails use generated msgID 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "partner-id:nbBDLgh4UgE7JJzqH/kxYjEnwuFWV010Imss5cLPqMU=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "GET",
      "params": Object {
        "currency": "VND",
        "grabID": "merchant-id",
        "msgID": "generated-msgID",
        "terminalID": "terminal-id",
        "txType": "P2M",
      },
      "url": "/mocapay/partners/v1/terminal/transaction/ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline getTxnDetails use provided msgID 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "partner-id:eU5VECl+LeHMdrmeTXJiTsTCLIqAiSNQxNro7+wpihY=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "GET",
      "params": Object {
        "currency": "VND",
        "grabID": "merchant-id",
        "msgID": "provided-msgID",
        "terminalID": "terminal-id",
        "txType": "P2M",
      },
      "url": "/mocapay/partners/v1/terminal/transaction/ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline performQrCode use generated msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"code\\":\\"code\\",\\"msgID\\":\\"generated-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:vCDyUDIlYVK7N4eTkSMwqFcfBoAXvtiSH/pWaws/Ljs=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partners/v1/terminal/transaction/perform",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline performQrCode use provided msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"code\\":\\"code\\",\\"msgID\\":\\"provided-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:MerEkT0zcq2u4p4QNQH6dBMReOb48vBsnzhhEiq/odA=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partners/v1/terminal/transaction/perform",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline refund use generated msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"amount\\":10000,\\"currency\\":\\"VND\\",\\"reason\\":\\"refund\\",\\"msgID\\":\\"generated-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:qMEueZX3ReliUUTa9L+iMVrGBH9BttgEpSCSx8syrwg=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "PUT",
      "url": "/mocapay/partners/v1/terminal/transaction/fBTcFk0T1OekPrLtbQoiIB33TyzwrWZl/refund",
    },
  ],
]
`;

exports[`MerchantIntegrationOffline refund use provided msgID 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"amount\\":10000,\\"partnerTxID\\":\\"ROe9fTsr8WYuyZHsuy738QXNtfyNyj9k\\",\\"currency\\":\\"VND\\",\\"reason\\":\\"refund\\",\\"msgID\\":\\"provided-msgID\\",\\"grabID\\":\\"merchant-id\\",\\"terminalID\\":\\"terminal-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:L2B5mWMD9ZzLI2Y2UeTFmjITVR0bjaKCCKGgjxY4fqY=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "PUT",
      "url": "/mocapay/partners/v1/terminal/transaction/fBTcFk0T1OekPrLtbQoiIB33TyzwrWZl/refund",
    },
  ],
]
`;
